DisCatSharp.ApplicationCommands/Attributes/SlashCommand/SlashCommandCooldownAttribute.cs

```diff
@@ -0,0 +1,305 @@
+// This file is part of the DisCatSharp project, based off DSharpPlus.
+//
+// Copyright (c) 2021-2022 AITSYS
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+using System;
+using System.Collections.Concurrent;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+
+using DisCatSharp.ApplicationCommands.Context;
+using DisCatSharp.ApplicationCommands.Enums;
+
+namespace DisCatSharp.ApplicationCommands.Attributes;
+
+/// <summary>
+/// Defines a cooldown for this command. This allows you to define how many times can users execute a specific command
+/// </summary>
+[AttributeUsage(AttributeTargets.Method | AttributeTargets.Class, AllowMultiple = true, Inherited = false)]
+public sealed class SlashCommandCooldownAttribute : SlashCheckBaseAttribute
+{
+	/// <summary>
+	/// Gets the maximum number of uses before this command triggers a cooldown for its bucket.
+	/// </summary>
+	public int MaxUses { get; }
+
+	/// <summary>
+	/// Gets the time after which the cooldown is reset.
+	/// </summary>
+	public TimeSpan Reset { get; }
+
+	/// <summary>
+	/// Gets the type of the cooldown bucket. This determines how cooldowns are applied.
+	/// </summary>
+	public CooldownBucketType BucketType { get; }
```

Comment:
I am noticing a pattern for these similar properties. This could be abstracted to an interface? Making the bucket feature a bit more robust since we can just reference the interface to do.... bucket things? Gotta pull branch and fiddle with this

Reply:
Thanks